### PR TITLE
[SPARK-27438][SQL] Parse strings with timestamps by to_timestamp() in microsecond precision

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

Add a `downScaleFactor` value to `UnixTime` to control scaling of returned values. Default value of `downScaleFactor` is `1000000` (microseconds per second) which means the result of string parsing in microseconds is divided by `1000000` in `UnixTimestamp` and `ToUnixTimestamp`. The `UnixTime` class is used in `ParseToTimestamp` which sets `downScaleFactor` to 1 as proposed in the PR. 

Similar `timestampScaleFactor` is added to the `Cast` expression to control scaling in casting `timestampScaleFactor ` * `long` <-> `timestamp`.

In this way, the `ParseToTimestamp` expression and `to_timestamp` function will be able to parse input strings with timestamps in microsecond precision.

## How was this patch tested?

By `DateFunctionsSuite`, and added new test to the test suite.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@srowen Please, take a look at the PR.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Review by srowen — APPROVED** [review]

**Comment by MaxGekk** [comment]
I will address test failures later.

**Comment by MaxGekk** [comment]
@srowen @cloud-fan Please, take a look at the PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -395,21 +399,21 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
...
   private[this] def decimalToTimestamp(d: Decimal): Long = {
-    (d.toBigDecimal * MICROS_PER_SECOND).longValue()
+    (d.toBigDecimal * timestampScaleFactor).longValue()
   }
   private[this] def doubleToTimestamp(d: Double): Any = {
-    if (d.isNaN || d.isInfinite) null else (d * MICROS_PER_SECOND).toLong
+    if (d.isNaN || d.isInfinite) null else (d * timestampScaleFactor).toLong
   }
 
   // converting seconds to us
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: we should update this comment

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 2]
```diff
@@ -395,21 +399,21 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
...
-  private[this] def longToTimestamp(t: Long): Long = SECONDS.toMicros(t)
+  private[this] def longToTimestamp(t: Long): Long = t * timestampScaleFactor
   // converting us to seconds
   private[this] def timestampToLong(ts: Long): Long = {
-    Math.floorDiv(ts, MICROS_PER_SECOND)
+    Math.floorDiv(ts, timestampScaleFactor)
   }
   // converting us to seconds in double
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: ditto

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 3]
```diff
@@ -203,7 +203,11 @@ object Cast {
       > SELECT _FUNC_('10' as int);
        10
   """)
-case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String] = None)
+case class Cast(
+    child: Expression,
+    dataType: DataType,
+    timeZoneId: Option[String] = None,
+    timestampScaleFactor: Long = MICROS_PER_SECOND)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: `Cast` is a widely used expression. Maybe a cleaner solution is to create a new expression `GetMicroSeconds` to convert a timestamp to microseconds count.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: I have to preserve semantic of `Cast` in `ParseToTimestamp`. I could create a parent class for `Cast` like `BaseCast` with `timestampScaleFactor`, and inherit it in `Cast` with `timestampScaleFactor` set to `MICROS_PER_SECOND`. WDYT?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: > I have to preserve semantic of Cast in ParseToTimestamp.

Can you elaborate on it? For end users, `ParseToTimestamp` and `Cast` are 2 function that are unrelated, though internally `ParseToTimestamp` is implemented via `Cast.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 4]
```diff
@@ -604,7 +604,11 @@ case class ToUnixTimestamp(
        1460041200
   """,
   since = "1.5.0")
-case class UnixTimestamp(timeExp: Expression, format: Expression, timeZoneId: Option[String] = None)
+case class UnixTimestamp(
+    timeExp: Expression,
+    format: Expression,
+    timeZoneId: Option[String] = None,
+    override val downScaleFactor: Long = MICROS_PER_SECOND)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: ditto, this violates the semantic of `UnixTimestamp`, which promises to return seconds.

We should create a new expression. We can create a base class for `UnixTimestamp` and the new expression to reuse code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: `ParseToTimestamp` extends `RuntimeReplaceable`,  and the optimizer replaces it by `Cast(UnixTimestamp)`. Not sure how it is important.

Frankly speaking, I would add new expression (and register a function) which just takes a string and a pattern, and parse it to `TimestampType` in microsecond precision. We have `DateTimeUtils.stringToTimestamp` but it doesn't accept a pattern.

Surprisingly, parsing of a string with specific pattern is possible by using `from_csv` now which not strait-forward method of doing this.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: >  For end users, ParseToTimestamp and Cast are 2 functions that are unrelated, though internally ParseToTimestamp is implemented via Cast.

`ParseToTimestamp` has to behave the same as `Cast` when a pattern is not provided.

**Comment by srowen** [comment]
@MaxGekk were you saying that you also thought this may not be worth the code and complication? I think that's a valid position too.

**Comment by MaxGekk** [comment]
> @MaxGekk were you saying that you also thought this may not be worth the code and complication?

Yes, I still think so. I would prefer to introduce a simple expression for parsing strings to timestamps in microsecond precision, and expose it as a function. It can be reused in another places if we will think it makes sense because the expression should just invoke `TimestampFormatter`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: we can do
```
def this(left: Expression, format: Expression) = {
    this(left, Option(format), GetMicroSeconds(UnixTimestamp(left, format))
  }

  def this(left: Expression) = this(left, None, Cast(left, TimestampType))
```

**Comment by cloud-fan** [comment]
> I would prefer to introduce a simple expression for parsing strings to timestamps in microsecond precision, and expose it as a function

We can keep the new expression internal, and change `ParseToTimestamp` to
```
  def this(left: Expression, format: Expression) = {
    this(left, Option(format), TheNewExpression(left, format))
  }

  def this(left: Expression) = this(left, None, Cast(left, TimestampType))
```

**Comment by MaxGekk** [comment]
I implemented an expression which just parses a string to timestamp, and used it in `ParseToTimestamp` as you pointed out above but one of tests failed here: https://github.com/apache/spark/blob/1470f23ec93f13d6c847832c06d40f8fc9803129/sql/core/src/test/scala/org/apache/spark/sql/DateFunctionsSuite.scala#L652-L653 . Should we support `to_timestamp` applied to non-string column with a pattern as second parameter? Does it make sense? Just in case, I tried to call different constructors using type of the first argument, but it doesn't work.

**Pull request closed without merging**